dagql/idtui/frontend_pretty.go

```diff
@@ -53,15 +53,19 @@ type frontendPretty struct {
 	cloudURL string
 
 	// TUI state/config
-	restore func()  // restore terminal
-	fps     float64 // frames per second
-	profile termenv.Profile
-	window  tea.WindowSizeMsg // set by BubbleTea
-	view    *strings.Builder  // rendered async
-	viewOut *termenv.Output
+	restore    func()  // restore terminal
+	fps        float64 // frames per second
+	profile    termenv.Profile
+	window     tea.WindowSizeMsg // set by BubbleTea
+	view       *strings.Builder  // rendered async
+	viewOut    *termenv.Output
+	browserBuf *strings.Builder // logs if browser fails
 
 	// held to synchronize tea.Model with updates
 	mu sync.Mutex
+
+	// messages to print before the final render
+	msgPreFinalRender strings.Builder
```

Comment:
nit: Seems simpler to just save a `loggedIn bool` alongside `cloudURL` and use that to conditionally print a message in `finalRender`, rather than keep an extra buffer

Reply:
> nit: Seems simpler to just save a `loggedIn bool` alongside `cloudURL` and use that to conditionally print a message in `finalRender`, rather than keep an extra buffer

I thought about it but I liked the fact that I could set the message in the `SetCloudURL` function and keep everything scoped there instead of having the logic split between two functions. Happy to add the isLogged flag is that seems like a better approach :pray:

Reply:
Ah I see the idea - sort of abstracting a 'pre' region for the final paint so that other places can write to it and keep all their concerns local. The cost of entry is still a code change in two spots, so the benefit comes in when multiple things need to populate it, but the issue I see there is we might actually care about the order that those separate pieces are rendered, which you have less control over it's all decoupled.

So in the end I think it's best to just push state into the model and let the render phase handle everything. TUI rendering is super fiddly so IME it's easier to treat it as one big monolith and pray it doesn't get too big. :sweat_smile: 

But yeah, really splitting hairs, if you agree feel free to make the change but no biggie either way

Reply:
hey @vito did not have the time to refactor this for the release. Will change in a follow-up PR :pray: